File: server/ops/diagnostics/checks/security/config-integrity.js
// Security check #1: has a secret/config file changed OUTSIDE anything this
// app itself did through a real route? Detection only, per the owner's own
// explicit scope — no access control, no prevention, nothing here ever
// blocks or reverts anything. Watches the files that actually matter if
// tampered with: .env (every API key), connections.json (model
// connections), external-services.json (TTS/STT keys), connectors.json
// (every connected app's own tokens).
//
// **Never fires on this app's own normal work, including a concurrent
// Claude session editing SOURCE — this only watches the four DATA/secret
// files above, never anything under server/ or public/.** Distinguishes
// "the app itself wrote this" from "something else did" using store.js's
// lastWriteAt()/config.js's envLastWriteTime() — an in-memory, per-process
// record of the exact CONTENT HASH the app itself last wrote, set at the
// ONE real write site each of these files has (writeJson()/writeEnvFile()).
//
// **A time-window version of this (compare against how RECENTLY the app
// wrote it, not WHAT it wrote) was tried first and found genuinely wrong by
// this check's own verification, not just reasoned about**: a legitimate
// write's timestamp stays "recent" long enough to wrongly excuse a LATER,
// unrelated external change landing inside the same window — confirmed
// live: writeJson() a real change, then immediately overwrite the same file
// externally (raw fs, bypassing store.js entirely) — the external change
// went undetected because the legitimate write's timestamp was still fresh.
// Comparing the actual hash the app produced instead means a legitimate
// write only ever explains the EXACT content it actually wrote, no matter
// how much time has passed; re-verified after the fix with the same
// reproduction, now correctly flagged.
//
// Keeps its own hash history via app_state (same pattern
// heartbeat/budget.js already uses), independent of the generic
// diagnostics engine's own per-run bookkeeping — this check owns its
// memory of "what did I see last time," never threaded through
// registry.js's own simple probe()/remedy() contract.

import fs from 'node:fs';
import crypto from 'node:crypto';
import { getDb } from '../../../../db.js';
import { dataFilePath, lastWriteAt } from '../../../../store.js';
import { envFilePath, envLastWriteTime } from '../../../../config.js';

const STATE_KEY = 'ops_diag_config_integrity';

const WATCHED = [
  { name: 'env', path: () => envFilePath(), lastAppWrite: () => envLastWriteTime() },
  { name: 'connections', path: () => dataFilePath('connections'), lastAppWrite: () => lastWriteAt('connections') },
  { name: 'external-services', path: () => dataFilePath('external-services'), lastAppWrite: () => lastWriteAt('external-services') },
  { name: 'connectors', path: () => dataFilePath('connectors'), lastAppWrite: () => lastWriteAt('connectors') },
];

function hashFile(filePath) {
  try {
    const buf = fs.readFileSync(filePath);
    return crypto.createHash('sha256').update(buf).digest('hex');
  } catch {
    return null; // doesn't exist yet — not a finding, just nothing to compare
  }
}

function readState() {
  const row = getDb().prepare('SELECT value FROM app_state WHERE key = ?').get(STATE_KEY);
  if (!row) return { hashes: {} };
  try {
    const parsed = JSON.parse(row.value);
    return parsed?.hashes ? parsed : { hashes: {} };
  } catch {
    return { hashes: {} };
  }
}

function writeState(state) {
  getDb()
    .prepare('INSERT INTO app_state (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value')
    .run(STATE_KEY, JSON.stringify(state));
}

export const id = 'security-config-integrity';

export async function probe() {
  const state = readState();
  const changed = [];
  const nextHashes = { ...state.hashes };

  for (const w of WATCHED) {
    const filePath = w.path();
    const currentHash = hashFile(filePath);
    const priorHash = Object.prototype.hasOwnProperty.call(state.hashes, w.name) ? state.hashes[w.name] : undefined;
    nextHashes[w.name] = currentHash;

    if (priorHash === undefined) continue; // first time seeing this file — establish baseline, not a finding
    if (currentHash === priorHash) continue; // unchanged (covers both non-null equal hashes and both-missing)

    // Changed. Explained only if the app's own last write to this file
    // produced EXACTLY this content — not merely "recently."
    const appWrite = w.lastAppWrite();
    const explainedByAppItself = appWrite?.hash === currentHash;
    if (!explainedByAppItself) changed.push(w.name);
  }

  writeState({ hashes: nextHashes });

  if (changed.length) {
    return { ok: false, detail: `Changed with no matching in-app write recorded: ${changed.join(', ')}.` };
  }
  return { ok: true };
}

// Deliberately NO remedy() — this build is detection only, per the owner's
// own explicit scope. A possible-tampering finding always escalates
// straight to the owner; nothing here ever reverts or blocks anything.

File: backend/jarvis/ops/diagnostics/checks/security/config_integrity.py
"""Did the secrets file change without this build having written it?

**Compared by the CONTENT HASH of what this app itself last wrote, not by how
recently it wrote something.** The time-window version of this check has a real
hole: any external change landing inside the same window is explained away by a
write it had nothing to do with. A hash cannot be fooled that way — either the
file is byte-for-byte what was last written, or something else changed it.

Never reports the file's contents, or any part of one. A check on a secrets file
that logs what it found would be the leak it exists to detect.
"""

from __future__ import annotations

import hashlib
from typing import Any

from .....db import get_db
from .....jscompat import now_iso
from ...registry import Check

STATE_KEY = "ops.config_integrity.last_seen"


def _hash_of(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _remembered() -> str | None:
    row = get_db().execute("SELECT value FROM app_state WHERE key = ?", (STATE_KEY,)).fetchone()
    return row["value"] if row is not None else None


def _remember(digest: str) -> None:
    get_db().execute("INSERT INTO app_state (key, value) VALUES (?, ?) "
                     "ON CONFLICT(key) DO UPDATE SET value = excluded.value",
                     (STATE_KEY, digest))


def probe() -> dict[str, Any]:
    from .....config import env_file_path, env_last_write_time

    path = env_file_path()
    if not path.exists():
        return {"ok": True}              # nothing configured yet is not a fault

    try:
        digest = _hash_of(path.read_text(encoding="utf-8"))
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "detail": f"The settings file could not be read: {err}"}

    own = (env_last_write_time() or {}).get("hash")
    if own == digest:
        _remember(digest)
        return {"ok": True}              # exactly what this build last wrote

    known = _remembered()
    if known is None:
        # First sight of the file this install — nothing to compare against, so
        # record it rather than reporting a change nobody can confirm happened.
        _remember(digest)
        return {"ok": True}
    if known == digest:
        return {"ok": True}              # unchanged since it was last seen

    _remember(digest)
    from . import counters
    counters.record("config_changed")
    return {"ok": False,
            "detail": ("The settings file holding API keys changed, and it was not this "
                       "app that changed it. Worth confirming that was you — "
                       f"noticed at {now_iso()}.")}


CHECK = Check(id="config_integrity", probe=probe, kind="security")
